docs/saved_objects/saved_object_repository_factory_design.md

```diff
@@ -0,0 +1,159 @@
+
+    * Pros: 
+        * No changes in core Dashboards. That means we can keep Dashboards as it is with very minimal changes.
+
+
+    * Cons
+        * Code reusability is low. 
+        * Some components of Saved object service such as Serializer, Type registry, interface to create internal and scoped repository are only available during Saved Object Service Start. As per the current architecture, first Saved Object Service Setup → Plugin Setup → Saved Object Service Start → Plugin Start. Some core plugin (e.g. opensearch_dashboards_usage_collection) calls find operation before plugin start and it fails because some components are still not available before plugin start.  
+
+- **Approach 3 (Preferred)**: The proposed architecture will add one more layer of abstraction in Saved Object Service. `The Repository Factory Provider` in Dashboards will be responsible for creating and managing instances of the Repository (e.g. SavedObjectRepository, PostgresRepository, DynamoDBRepository etc.), which is used to interact with the metadata storage that stores the saved objects. Currently we have an repository interface named `ISavedObjectRepository`, and the `SavedObjectRepository` is the implementation, which use an OpenSearch index as its data store. This approach would make the implementation of `ISavedObjectRepository` replaceable by plugin.
+
+    ![img](./img/proposed_saved_object_service_workflow.png)
+
+   * Pros: 
+        * Only change needed in Dashboard is to introduce one more abstraction layer in Saved Object Service.
+        * Adds opportunity for community developers to contribute for other meta store.
+        
+    * Cons
+        * Code reusability is low.  
+
+`SavedObjectsServiceSetup` provides interface to create custom Saved Object Repository.
+```
+/**
+* Set the default {@link SavedObjectRepositoryFactoryProvider | factory provider} for creating Saved Objects repository.
+* Only one repository can be set, subsequent calls to this method will fail.
+*/
+registerRepositoryFactoryProvider: (
+respositoryFactoryProvider: SavedObjectRepositoryFactoryProvider
+) => void;
+```
+
+Here are the main steps involved in using the Saved Objects Repository Factory in Dashboards:  
+1. Define the dependencies: The Saved Object Repository Factory Provider requires the function which creates instance of `ISavedObjectRepository`.
+    ```
+    export const repositoryFactoryProvider: SavedObjectRepositoryFactoryProvider = (
+    options: SavedObjectsRepositoryOptions
+    ) => {
+    .
+    .
+    .
+    return new PostgresRepository({
+        typeRegistry,
+        serializer,
+        migrator,
+        allowedTypes,
+    });
+    }
+    ```
+2. Register the provider: Register the repository factory provider with right dependencies. 
+    ```
+    core.savedObjects.registerRepositoryFactoryProvider(repositoryFactoryProvider);
```

Comment:
use `portgresRepositoryFactory` for naming to better callout the usage?

Reply:
`Repository factory provider` is not limited to Postgres. PostgresRepository is just one use case.